bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -0,0 +1,551 @@
+        BigtableTableName tableName = clusterName.toTableName(configuration.getTableId());
+        SampleRowKeysRequest.Builder request = SampleRowKeysRequest.newBuilder();
+        request.setTableName(tableName.toString());
+
+        List<SampleRowKeysResponse> sampleRowKeys =
+            session.getDataClient().sampleRowKeys(request.build());
+        byte[] startKey = HConstants.EMPTY_START_ROW;
+        long lastOffset = 0;
+        for (SampleRowKeysResponse response : sampleRowKeys) {
+          byte[] endKey = response.getRowKey().toByteArray();
+          // Avoid empty regions.
+          if (Bytes.equals(startKey, endKey) && startKey.length > 0) {
+            continue;
+          }
+
+          long offset = response.getOffsetBytes();
+          // Get all the start/end key ranges that match the user supplied Scan.  See
+          // https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/TableInputFormatBase.java#L298
+          // for more info.
+          if ((scanStartKey.length == 0 || endKey.length == 0
+                  || Bytes.compareTo(scanStartKey, endKey) < 0)
+              && (scanEndKey.length == 0 || Bytes.compareTo(scanEndKey, startKey) > 0)) {
+            byte[] splitStart =
+                scanStartKey.length == 0 || Bytes.compareTo(startKey, scanStartKey) >= 0
+                    ? startKey : scanStartKey;
+            byte[] splitStop =
+                (scanEndKey.length == 0 || Bytes.compareTo(endKey, scanEndKey) <= 0)
+                    && endKey.length > 0
+                    ? endKey
+                    : scanEndKey;
+            split(splits, offset - lastOffset, desiredBundleSizeBytes, splitStart, splitStop);
+          }
+          lastOffset = offset;
+          startKey = endKey;
+        }
+        // Create one last region if the last region doesn't reach the end or there are no regions.
+        byte[] endKey = HConstants.EMPTY_END_ROW;
+        if (!Bytes.equals(startKey, endKey)) {
+          splits.add(new SourceWithKeys(startKey, endKey, 0));
+        }
+      }
+      return splits;
+    }
+
+    @Override
+    public void validate() {
+      checkNotNullOrEmpty(configuration.getProjectId(), "project");
+      checkNotNullOrEmpty(configuration.getZoneId(), "zone");
+      checkNotNullOrEmpty(configuration.getClusterId(), "cluster");
+      checkNotNullOrEmpty(configuration.getTableId(), "table");
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+      BigtableOptions bigtableOptions = configuration.toBigtableOptions();
+      long totalEstimatedSizeBytes = 0;
+
+      Scan scan = configuration.getScan();
+      byte[] scanStartKey = scan.getStartRow();
+      byte[] scanEndKey = scan.getStopRow();
+
+      List<SourceWithKeys> splits = new ArrayList<>();
+      try (BigtableSession session = new BigtableSession(bigtableOptions)) {
```

Comment:
Make this a separate function returning  List<SampleRowKeysResponse>?

+    try (BigtableSession session = new BigtableSession(bigtableOptions)) {
+        BigtableClusterName clusterName = bigtableOptions.getClusterName();
+        BigtableTableName tableName = clusterName.toTableName(configuration.getTableId());
+        SampleRowKeysRequest.Builder request = SampleRowKeysRequest.newBuilder();
+        request.setTableName(tableName.toString());
+        List<SampleRowKeysResponse> sampleRowKeys =
+            session.getDataClient().sampleRowKeys(request.build());